azavea/iow-boundary-tool
# Create Remaining Initial Django Models

## Overview

This creates the initial Django models needed for the different states of submission as well as for the starting user interactions on the frontend, as discussed in the issue comments.

Closes #69 

### Demo

An [ADR](doc/arch/adr-001-data-models.md) is available as part of the PR.

<img width="683" alt="Screen Shot 2022-09-28 at 9 20 11 PM" src="https://user-images.githubusercontent.com/36374797/192917000-90d2de7c-6035-4e0b-a2d1-75ea91f632cf.png">

### Notes

- The `State` default instance, North Carolina, is created through a data migration that references a GEOJSON file downloaded during `scripts/fetch-data`. I found the state boundary url [here](https://public.opendatasoft.com/explore/dataset/us-state-boundaries/export/?refine.name=North+Carolina), but not sure if there's a better resource for state boundary geojson files? I also included `scripts/fetch-data` as part of the setup script so that the GEOJSON gets setup before running migrations. I didn't see any downsides to this, but not sure if I'm missing context on fetching the municipality data (the original script purpose)?
- I added `__str__` methods to some of the models where it felt appropriate, but not all. I wasn't sure what language to use to best describe instances like the "approval of the submission of a boundary of a utility" when there wasn't a unique identifier besides the utility, and it seemed to better fit a future task once we better know use case anyways.
- I decided to group together `Submission`, `Approval`, `Review`, and `Annotation` models into one file. These felt like all related pieces of the submission process that could benefit from being organized together, but curious what others think here.
- ~~I made the `max_length` of the `Submission.upload_filename` and `ReferenceImage.filename` fields match the `max_length` of the [Django `FileField`](https://docs.djangoproject.com/en/4.1/ref/models/fields/#filefield) since that seemed like an appropriate reference.~~ These fields' `max_length` are now `255`.


## Testing Instructions

- `./scripts/fetch-data`
- `./scripts/update`
- `./scripts/server`
- Visit the below Django admin pages for each newly created object:
  - [State](http://localhost:8181/admin/api/state/)
  - [Boundary](http://localhost:8181/admin/api/boundary/)
  - [Submission](http://localhost:8181/admin/api/submission/)
  - [ReferenceImage](http://localhost:8181/admin/api/referenceimage/)
  - [Review](http://localhost:8181/admin/api/review/)
  - [Annotation](http://localhost:8181/admin/api/annotation/)
  - [Approval](http://localhost:8181/admin/api/approval/)
- Confirm the correct field and field types for each object
- Confirm correct required field validation on save
- Confirm North Carolina is a created `State` instance and fields are as expected
- Confirm `User` and `Utility` have new `state` fields
- Confirm `Utility`'s `state` field defaults to North Carolina

## Checklist
